File: src/lib/billable-engine.ts
import billableRulesConfig from "@/config/billable-rules.json";
import type {
  BillableSource,
  BillableStatus,
  ExtractedReceipt,
  ExpenseCategory,
  ScannedReceipt,
} from "@/lib/types";

type RuleType =
  | "merchant_blocklist"
  | "merchant_allowlist"
  | "category"
  | "amount_over"
  | "line_item_keyword";

interface BillableRule {
  id: string;
  type: RuleType;
  patterns?: string[];
  categories?: ExpenseCategory[];
  category?: ExpenseCategory;
  maxAmount?: number;
  keywords?: string[];
  billableStatus: BillableStatus;
  reason: string;
}

interface BillableRulesConfig {
  rules: BillableRule[];
  default: {
    billableStatus: BillableStatus;
    reason: string;
  };
}

export interface BillableEvaluation {
  billableStatus: BillableStatus;
  billableReason: string;
  billableSource: BillableSource;
  matchedRuleId?: string;
}

const config = billableRulesConfig as BillableRulesConfig;
const WORK_ORDER_BILLABLE_REASON =
  "Work order attached — receipts with work orders are always billable.";

function normalizeText(value: string): string {
  return value.toLowerCase().trim();
}

function merchantMatches(merchant: string, patterns: string[]): boolean {
  const normalizedMerchant = normalizeText(merchant);
  return patterns.some((pattern) =>
    normalizedMerchant.includes(normalizeText(pattern)),
  );
}

function lineItemsText(receipt: ExtractedReceipt): string {
  return receipt.lineItems.map((item) => normalizeText(item.name)).join(" ");
}

function evaluateRule(
  rule: BillableRule,
  receipt: ExtractedReceipt,
): boolean {
  switch (rule.type) {
    case "merchant_blocklist":
    case "merchant_allowlist":
      return merchantMatches(receipt.merchant, rule.patterns ?? []);
    case "category":
      return (rule.categories ?? []).includes(receipt.category);
    case "amount_over":
      return (
        receipt.category === rule.category &&
        typeof rule.maxAmount === "number" &&
        receipt.amount > rule.maxAmount
      );
    case "line_item_keyword": {
      const haystack = `${normalizeText(receipt.merchant)} ${lineItemsText(receipt)}`;
      return (rule.keywords ?? []).some((keyword) =>
        haystack.includes(normalizeText(keyword)),
      );
    }
    default:
      return false;
  }
}

interface EvaluationPhase {
  types: RuleType[];
  statuses?: BillableStatus[];
}

function findMatchingRule(
  rules: BillableRule[],
  receipt: ExtractedReceipt,
  phase: EvaluationPhase,
): BillableRule | undefined {
  return rules.find(
    (rule) =>
      phase.types.includes(rule.type) &&
      (!phase.statuses || phase.statuses.includes(rule.billableStatus)) &&
      evaluateRule(rule, receipt),
  );
}

/** Non-billable checks run before allowlists so item context wins over store name. */
const EVALUATION_PHASES: EvaluationPhase[] = [
  { types: ["merchant_blocklist"] },
  { types: ["line_item_keyword"], statuses: ["non_billable"] },
  { types: ["category"], statuses: ["non_billable"] },
  { types: ["merchant_allowlist"] },
  { types: ["amount_over"] },
  { types: ["line_item_keyword"], statuses: ["billable", "review"] },
  { types: ["category"], statuses: ["billable", "review"] },
];

export function evaluateBillable(
  receipt: ExtractedReceipt,
  rules: BillableRulesConfig = config,
): BillableEvaluation {
  if (receipt.workOrderNumber) {
    return {
      billableStatus: "billable",
      billableReason: WORK_ORDER_BILLABLE_REASON,
      billableSource: "rule",
      matchedRuleId: "work-order-attached",
    };
  }

  for (const phase of EVALUATION_PHASES) {
    const matchingRule = findMatchingRule(rules.rules, receipt, phase);

    if (matchingRule) {
      return {
        billableStatus: matchingRule.billableStatus,
        billableReason: matchingRule.reason,
        billableSource: "rule",
        matchedRuleId: matchingRule.id,
      };
    }
  }

  return {
    billableStatus: rules.default.billableStatus,
    billableReason: rules.default.reason,
    billableSource: "default",
  };
}

export function applyBillableEvaluation(
  receipt: ExtractedReceipt,
  evaluation: BillableEvaluation,
): ScannedReceipt {
  return {
    ...receipt,
    ...evaluation,
  };
}

export function evaluateAndApplyBillable(
  receipt: ExtractedReceipt,
): ScannedReceipt {
  return applyBillableEvaluation(receipt, evaluateBillable(receipt));
}

export function normalizeBillableFields(
  expense: Partial<ScannedReceipt>,
): Pick<
  ScannedReceipt,
  "billableStatus" | "billableReason" | "billableSource" | "matchedRuleId"
> {
  if (expense.workOrderNumber) {
    return {
      billableStatus: "billable",
      billableReason: WORK_ORDER_BILLABLE_REASON,
      billableSource: "rule",
      matchedRuleId: "work-order-attached",
    };
  }

  if (
    expense.billableStatus &&
    expense.billableReason &&
    expense.billableSource
  ) {
    return {
      billableStatus: expense.billableStatus,
      billableReason: expense.billableReason,
      billableSource: expense.billableSource,
      matchedRuleId: expense.matchedRuleId,
    };
  }

  return {
    billableStatus: "review",
    billableReason: "Legacy receipt — re-scan or review manually",
    billableSource: "default",
  };
}

File: agent/src/kai_agent/services/billable_engine.py
"""Billable rules engine — port of src/lib/billable-engine.ts using shared JSON config."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Literal

BillableStatus = Literal["billable", "non_billable", "review"]
RuleType = Literal[
    "merchant_blocklist",
    "merchant_allowlist",
    "category",
    "amount_over",
    "line_item_keyword",
]

EVALUATION_PHASES: list[dict[str, Any]] = [
    {"types": ["merchant_blocklist"]},
    {"types": ["line_item_keyword"], "statuses": ["non_billable"]},
    {"types": ["category"], "statuses": ["non_billable"]},
    {"types": ["merchant_allowlist"]},
    {"types": ["amount_over"]},
    {"types": ["line_item_keyword"], "statuses": ["billable", "review"]},
    {"types": ["category"], "statuses": ["billable", "review"]},
]


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[4]


def load_billable_rules() -> dict[str, Any]:
    path = _repo_root() / "src" / "config" / "billable-rules.json"
    with path.open(encoding="utf-8") as f:
        return json.load(f)


def _normalize_text(value: str) -> str:
    return value.lower().strip()


def _merchant_matches(merchant: str, patterns: list[str]) -> bool:
    normalized = _normalize_text(merchant)
    return any(_normalize_text(pattern) in normalized for pattern in patterns)


def _line_items_text(receipt: dict[str, Any]) -> str:
    items = receipt.get("line_items") or receipt.get("lineItems") or []
    return " ".join(_normalize_text(str(item.get("name", ""))) for item in items)


def _evaluate_rule(rule: dict[str, Any], receipt: dict[str, Any]) -> bool:
    rule_type = rule["type"]
    merchant = str(receipt.get("merchant") or receipt.get("vendor") or "")

    if rule_type in ("merchant_blocklist", "merchant_allowlist"):
        return _merchant_matches(merchant, rule.get("patterns") or [])

    if rule_type == "category":
        return receipt.get("category") in (rule.get("categories") or [])

    if rule_type == "amount_over":
        max_amount = rule.get("maxAmount")
        return (
            receipt.get("category") == rule.get("category")
            and isinstance(max_amount, (int, float))
            and float(receipt.get("amount") or 0) > float(max_amount)
        )

    if rule_type == "line_item_keyword":
        haystack = f"{_normalize_text(merchant)} {_line_items_text(receipt)}"
        return any(
            _normalize_text(keyword) in haystack for keyword in (rule.get("keywords") or [])
        )

    return False


def _find_matching_rule(
    rules: list[dict[str, Any]],
    receipt: dict[str, Any],
    phase: dict[str, Any],
) -> dict[str, Any] | None:
    phase_types: list[str] = phase["types"]
    phase_statuses = phase.get("statuses")

    for rule in rules:
        if rule["type"] not in phase_types:
            continue
        if phase_statuses and rule.get("billableStatus") not in phase_statuses:
            continue
        if _evaluate_rule(rule, receipt):
            return rule

    return None


def evaluate_billable(
    receipt: dict[str, Any],
    config: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Return billableStatus, billableReason, matchedRuleId."""
    rules_config = config or load_billable_rules()
    rules: list[dict[str, Any]] = rules_config.get("rules") or []

    for phase in EVALUATION_PHASES:
        match = _find_matching_rule(rules, receipt, phase)
        if match:
            return {
                "billableStatus": match["billableStatus"],
                "billableReason": match["reason"],
                "matchedRuleId": match["id"],
            }

    default = rules_config.get("default") or {}
    return {
        "billableStatus": default.get("billableStatus", "review"),
        "billableReason": default.get(
            "reason",
            "Not clearly a property or client expense — needs manual review",
        ),
        "matchedRuleId": None,
    }
